expand.rkt

```diff
@@ -0,0 +1,97 @@
+#lang racket/base
+
+(provide read-and-expand)
+
+(require racket/match
+         racket/async-channel
+         syntax/modread
+         syntax/parse
+         drracket/check-syntax)
+
+(define *expander-ch* (make-async-channel))
+
+;; (-> input-port path collector% async-channel)
+;; async-channel receives (list (or syntax #f) (or expand-syntax #f))
+;; where `syntax` is the result of read-syntax,
+;; `expand-syntax` is the result of expand.
+;; They can be #f if error happens during processing.
+(define (read-and-expand in path collector)
+  (define ch (make-async-channel))
+  (async-channel-put *expander-ch* (list path in ch collector))
+  ch)
+
+(define (expander)
+  ;; cache namespace for continuous expand request for same uri
```

Comment:
Can we cache the namespace via path? i.e. A hashmap that

path1 |-> ns1
path2 |-> ns2
...

Reply:
It's kind of difficult.

The cached namespace would be invalid if the dependency module changes. There also should be a expiration mechanism to remove cache and avoid memory leak.

But it's tricky to know when it would be invalid and when it becomes useless so we can remove it.

Reply:
I see, so unless we have a full dependencies tracker system (and must integrate with editor changes), we don't have easy way to find a namespace is invalid. Then the current one is acceptable, thanks!